text/3691-trusted-publishing-cratesio.md

```diff
@@ -0,0 +1,371 @@
+3. After authenticating the workflow's request for an ID token, the GitHub Actions OIDC provider returns a cryptographically signed ID token (signed by its private key) .
+4. The release workflow, sends the signed ID token to crates.io to exchange for an access token.
+    - The ID token is not used to authorize any action on crates.io, it's only used to authenticate.
+5. Crates.io, using the public keys of the GitHub Action OIDC Provider, validates the signature of the provided ID token that has been signed.
+6. Once validated, crates.io generates a temporary access token and returns it to the Github Actions release workflow.
+7. The release workflow builds the artifact to be published (i.e. [the `.crate` file](https://doc.rust-lang.org/cargo/commands/cargo-package.html))
+8. The release workflow, using the access token retrieved from step 6, authenticates to crates.io and publishes the version of the crate.
+9. As a good steward, once the crate has successfully been published, the workflow job will request the revocation of the access token from crates.io
+    - An access token's lifetime might outlive the actual time needed to perform the action. It's good practice to revoke it once done to further limit any risks of accidental exposure. 
+
+> [!NOTE]
+> Steps 7 and 8 are abstracted in the `cargo publish` command. The access token would be passed via the `CARGO_REGISTRY_TOKEN` environment variable.
+ 
+</details>
+
+### Trusted Publisher Configuration on crates.io
+[trusted-publisher-configuration-cratesio]: #trusted-publisher-configuration-cratesio
+
+Prior to publishing a crate through a GitHub Actions workflow, the crate author will need to create a _Trusted Publisher Configuration_ via the crates.io web UI. This is a one-time configuration **required per crate**.
+
+> [!Note]
+> Initially, a _Trusted Publisher Configuration_ can only be created after an initial manual publishing of a crate.
+> 
+> See [Future Possibilities](#future-possibilities) for future plans to alleviate this pain.
+
+Without a corresponding _Trusted Publisher Configuration_ any attempts from a GitHub Action workflow to exchange an ID token for an access token from crates.io will be unauthorized and fail. 
+
+A _Trusted Publisher Configuration_ will vary on requirements depending on the trusted publisher being used (e.g. GitHub Actions, GitLab Pipelines, and so on). CI/CD providers will vary in the claims they make within their ID token. 
+
+For GitHub Actions, the crate author will need to provide:
+
+1. (required) The owning GitHub username or organization
+2. (required) The repository name
+3. (required) The workflow file name (must be located in `.github/workflows/`)
+4. (optional) The [GitHub Actions environment](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment) name
```

Comment:
On PyPI we treat "no environment specified" as "any environment is allowed," since we observed that users _would_ configure an environment on the CI/CD side and then forget to include it on the Trusted Publishing side. They'd then get confused as to why their publisher wasn't working, since the error would strongly suggest that their configuration was correct (the only difference being the environment).

That being said, YMMV! PyPI's approach was pretty experimental, and it's possible that you can avoid this user confusion through better documentation and error messaging.

Reply:
> They'd then get confused as to why their publisher wasn't working, since the error would strongly suggest that their configuration was correct (the only difference being the environment).

Would it have been possible to in the error message suggest to change the publish settings on the PyPI side to allow the environment that was used?

Reply:
> Would it have been possible to in the error message suggest to change the publish settings on the PyPI side to allow the environment that was used?

Yep, that's pretty much what we do now -- we didn't do that originally because we were wary of users blindly trusting the error message to tell them what to configure, but in practice I think it's a reasonable tradeoff

Reply:
I'm not sure I have a strong opinion on this either way. Environments are a newish feature on GitHub Actions, and I generally don't see them used much. Treating the empty `environment` field as an _any_ felt nature to me.

GitHub Action Environments are generally used to increase the security around a workflow, so it feels weird to explicitly not want one. 🤔